wasi/core/src/swam/runtime/wasi/Core.scala

```diff
@@ -0,0 +1,386 @@
+    (for {
+      _ <- mem.writeInt(argc, args.length)
+      _ <- mem.writeInt(argvBufSize, args.map(t => t.length + 1).sum)
+    } yield success)
+      .handleErrorWith(t => logger.error("Could not get the argument sizes", t).as(io))
+
+  def environGet(environ: Pointer, buf: Pointer): F[Errno] =
+    System
+      .getenv()
+      .asScala
+      .toList
+      .foldLeftM((environ, buf)) {
+        case ((coffset, offset), (name, value)) =>
+          val bytes = s"$name=$value".getBytes()
+          for {
+            _ <- mem.writeInt(coffset, offset)
+            _ <- mem.writeBytes(offset, bytes)
+            _ <- mem.writeByte(offset + bytes.length, 0)
+          } yield (coffset + 4, offset + bytes.length + 1)
+      }
+      .as(success)
+      .handleErrorWith(t => logger.error("Could not get the environment", t).as(io))
+
+  def environSizesGet(environc: Pointer, environBufSize: Pointer): F[Errno] = {
+    val (size, count) =
+      System
+        .getenv()
+        .asScala
+        .foldLeft((0, 0)) {
+          case ((cumul, count), (name, value)) =>
+            val entrySize = name.length + value.length + 2
+            (cumul + entrySize, count + 1)
+        }
+
+    (for {
+      _ <- mem.writeInt(environc, count)
+      _ <- mem.writeInt(environBufSize, size)
+    } yield success)
+      .handleErrorWith(t => logger.error("Could not get the environment size", t).as(io))
+  }
+
+  // TODO I assume nanosecond precision is available but it depends on the `Clock` implementation
+  // and the underlying platform
+  def clockResGet(id: Clockid, resolution: Pointer): F[Errno] =
+    id match {
+      case Monotonic =>
+        mem
+          .writeLong(resolution, 1L)
+          .as(success)
+          .handleErrorWith(t => logger.error("Could not get the clock resolution", t).as(io))
+      case Realtime =>
+        mem
+          .writeLong(resolution, 1000000L)
+          .as(success)
+          .handleErrorWith(t => logger.error("Could not get the clock resolution", t).as(io))
+      case _ =>
+        F.pure(inval)
+    }
+
+  def clockTimeGet(id: Clockid, precision: Timestamp, time: Pointer): F[Errno] =
+    id match {
+      case Monotonic =>
+        (for {
+          nanos <- clock.monotonic(TimeUnit.NANOSECONDS)
+          _ <- mem.writeLong(time, nanos)
+        } yield success)
+          .handleErrorWith(t => logger.error("Could not get the clock value", t).as(io))
+      case Realtime =>
+        (for {
+          millis <- clock.realTime(TimeUnit.MILLISECONDS)
+          _ <- mem.writeLong(time, millis)
+        } yield success)
+          .handleErrorWith(t => logger.error("Could not get the clock value", t).as(io))
+      case _ =>
+        F.pure(inval)
+    }
+
+  def fdAdvise(fd: Fd, offset: Filesize, len: Filesize, advice: Advice): F[Errno] =
+    unimplemented("fd_advise")
+
+  def fdAllocate(fd: Fd, offset: Filesize, len: Filesize): F[Errno] =
+    unimplemented("fs_allocate")
+
+  def pathOpen(fd: Fd,
```

Comment:
@ybasket Just as a sanity check, could you please have a look at my use of cats-effect in this method, and more generally this class? :pray: